drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1931 @@
+	__ASSERT_NO_MSG(msg->msg_type == NUMAKER_USBD_MSG_TYPE_SW_RECONN);
+
+	/* S/W reconnect for error recovery */
+	numaker_usbd_lock(dev);
+	numaker_usbd_sw_reconnect(dev);
+	numaker_usbd_unlock(dev);
+}
+
+/* Message handler for callback for usb_dc_status_code */
+static void numaker_usbd_msg_cb_state(struct usb_dc_numaker_device *dev,
+				      struct numaker_usbd_msg *msg)
+{
+	struct usb_dc_numaker_data *data = dev->data;
+
+	__ASSERT_NO_MSG(msg->msg_type == NUMAKER_USBD_MSG_TYPE_CB_STATE);
+
+	/* Interrupt bottom half processing for bus reset */
+	if (msg->cb_state.status_code == USB_DC_RESET) {
+		numaker_usbd_lock(dev);
+		numaker_usbd_bus_reset_bh(dev);
+		numaker_usbd_unlock(dev);
+	}
+
+	/* NOTE: Don't run callback with our mutex locked, or we may encounter
+	 * deadlock because the Zephyr USB device stack can have its own
+	 * synchronization.
+	 */
+	if (data->status_cb) {
+		data->status_cb(msg->cb_state.status_code, NULL);
+	} else {
+		LOG_WRN("No status callback: status_code=%d", msg->cb_state.status_code);
+	}
+}
+
+/* Message handler for callback for usb_dc_ep_cb_status_code */
+static void numaker_usbd_msg_cb_ep(struct usb_dc_numaker_device *dev, struct numaker_usbd_msg *msg)
+{
+	__ASSERT_NO_MSG(msg->msg_type == NUMAKER_USBD_MSG_TYPE_CB_EP);
+
+	uint8_t ep_addr = msg->cb_ep.ep_addr;
+
+	/* Bind EP context to EP address */
+	struct numaker_usbd_ep *ep_cur = numaker_usbd_ep_mgmt_bind_ep(dev, ep_addr);
+
+	if (!ep_cur) {
+		LOG_ERR("Bind EP context: ep_addr=0x%02x", ep_addr);
+		return;
+	}
+
+	/* Interrupt bottom half processing for EP */
+	numaker_usbd_lock(dev);
+	numaker_usbd_ep_bh(ep_cur, msg->cb_ep.status_code);
+	numaker_usbd_unlock(dev);
+
+	/* NOTE: Same as above, don't run callback with our mutex locked */
+	if (ep_cur->cb) {
+		ep_cur->cb(ep_addr, msg->cb_ep.status_code);
+	} else {
+		LOG_WRN("No EP callback: ep_addr=0x%02x, status_code=%d", ep_addr,
+			msg->cb_ep.status_code);
+	}
+}
+
+/* Interrupt bottom half processing
+ *
+ * This thread is used to not run Zephyr USB device stack and callbacks in interrupt
+ * context. This is because callbacks from this stack may use mutex or other kernel functions
+ * which are not supported in interrupt context.
+ */
+static void numaker_usbd_msg_hdlr_thread_main(void *arg1, void *arg2, void *arg3)
+{
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	__ASSERT_NO_MSG(arg1);
+	struct usb_dc_numaker_device *dev = (struct usb_dc_numaker_device *)arg1;
+	struct usb_dc_numaker_data *data = dev->data;
+
+	struct numaker_usbd_msg msg;
+
+	while (true) {
+		if (k_msgq_get(&data->msgq, &msg, K_FOREVER)) {
+			continue;
+		}
+
+		switch (msg.msg_type) {
+		case NUMAKER_USBD_MSG_TYPE_SW_RECONN:
+			numaker_usbd_msg_sw_reconn(dev, &msg);
+			break;
+
+		case NUMAKER_USBD_MSG_TYPE_CB_STATE:
+			numaker_usbd_msg_cb_state(dev, &msg);
+			break;
+
+		case NUMAKER_USBD_MSG_TYPE_CB_EP:
+			numaker_usbd_msg_cb_ep(dev, &msg);
+			break;
+
+		default:
+			__ASSERT_NO_MSG(false);
+		}
+	}
+}
+
+/* ISR */
```

Comment:
ok, it's in the function name, just drop it

Reply:
dropped